File: src/lib/categorize.ts
import type { CategoryRule } from '../db/schema';

export function suggestCategory(
  description: string,
  rules: CategoryRule[],
): string | null {
  const d = description.toLowerCase();
  let best: CategoryRule | null = null;
  for (const r of rules) {
    if (!d.includes(r.keyword.toLowerCase())) continue;
    if (
      !best ||
      r.priority > best.priority ||
      (r.priority === best.priority && r.keyword.length > best.keyword.length)
    ) {
      best = r;
    }
  }
  return best?.categoryId ?? null;
}

File: lib/categorize.dart
import 'package:centavoo/models/category_rule.dart';

String? suggestCategory(String description, List<CategoryRule> rules) {
  final d = description.toLowerCase();
  CategoryRule? best;
  for (final r in rules) {
    if (!d.contains(r.keyword.toLowerCase())) continue;
    if (best == null ||
        r.priority > best.priority ||
        (r.priority == best.priority && r.keyword.length > best.keyword.length)) {
      best = r;
    }
  }
  return best?.categoryId;
}
